### PR TITLE
Migrate `logits` computation and gather to `model_runner`

### PR DESCRIPTION
**Pull request by esmeetu**
1. Finish TODO: https://github.com/vllm-project/vllm/pull/2221#discussion_r1437310420
2. ~~This might replace #2398 .~~   (refer to #3500)

### Description
This PR migrate `logits` computation and gather to `model_runner`. This change make `Sampler` simple and clean.
Furthermore, i want to remove `sample` method from model file, like `llama.py`. Because `model` and `sample` are at different stage, and should decouple with each other. But in this comment of PR #3183 : https://github.com/vllm-project/vllm/pull/3183#discussion_r1513815746 , that model will scale `logits` in sampler, so i will keep `sample` method as it is. This PR will better support #3183 model integration as well.

### Pipeline:

1. Prepare inputs
1. Model
2. **LogitsProcessor**  (support logits scale and Custom LogitsProcessor Functions)
3. Sampler
4. Output

### TODO

- [x] Fix llama, opt for tests
- [x] Fix tests (sampler and logits_processor)
- [x] Fix other model files
~~- [ ] Fix #2398~~

### CODE REVIEW

**Comment by grandiose-pizza** [comment]
Hi @esmeetu , Important update.

Regarding this comment, https://github.com/vllm-project/vllm/pull/3183#issuecomment-1980110711

I found the bug. It will be useful for this PR.

Please add this change while making the PR for scaling logits: https://github.com/vllm-project/vllm/pull/3183/commits/e04e56d981bc3fc81f94307c5c274ed065a13bd3

This is done as during multiGPU setting logits are None for `get_tensor_model_parallel_rank()` that are > 1 (it will be 0 & 1 in a 2GPU setting) as per this:
https://github.com/vllm-project/vllm/blob/a33ce60c6629e8c22aaf002ae8478a685e726e3e/vllm/model_executor/parallel_utils/communication_op.py#L102

**Comment by Yard1** [comment]
While I agree with this change in principle, I think it's important to ensure we have an API that can support different usecases. For example, I would suggest making the logit generation a layer (or some other sort of abstraction on the model level). The fact that it has now been moved out of the model into the model runner makes the code harder to understand, especially considering the sampler remains a layer.

In other words, I would just suggest adding a new logit generator layer to the models (or the sampler, though models would be better - the output of a model should be logits IMO) and not putting that logic inside of model runner.

**Comment by esmeetu** [comment]
@Yard1 @zhuohan123 I redesign this PR, which make logits processor a individual layer. PTAL!
After this design is supported, i will fix other model files.

**Comment by Yard1** [comment]
This need changes to work with the LoRA path

**Comment by esmeetu** [comment]
@Yard1 All CI passed, please review this again. cc @zhuohan123 
And should we resolve #2398 into this PR? Is there any better solution?

File: vllm/worker/model_runner.py [hunk 1]
```diff
@@ -576,9 +576,15 @@ def execute_model(
             input_metadata=input_metadata,
         )
 
+        # Compute the logits.
+        logits = self.model.compute_logits(hidden_states, sampling_metadata)
+
+        if not sampling_metadata.perform_sampling:
```
Comment (zhuohan123) [inline vllm/worker/model_runner.py:582]:
```suggestion
        # Only perform sampling in the driver worker.
        if not sampling_metadata.perform_sampling:
```

**Review by zhuohan123 — APPROVED** [review]
Thanks for the fix! This looks very good. Much better than what we have before :) Can you merge the PR after fixing the merge conflicts?
